caffe2/operators/tensorrt_op_trt.cc

```diff
@@ -0,0 +1,172 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "caffe2/operators/tensorrt_op_trt.h"
+
+#include <unordered_map>
+
+namespace caffe2 {
+
+namespace {
+struct InferDeleter {
+  template <typename T>
+  void operator()(T* obj) const {
+    if (obj) {
+      obj->destroy();
+    }
+  }
+};
+
+template <typename T>
+inline std::shared_ptr<T> InferObject(T* obj) {
+  if (!obj) {
+    CAFFE_THROW("Failed to create TensorRt object");
+  }
+  return std::shared_ptr<T>(obj, InferDeleter());
+}
+
+// Note that input of trt tensor is in CHW format, while our tensor is NCHW
+bool CheckDims(const nvinfer1::Dims& nv_dims, const std::vector<TIndex>& c2_dims) {
+  if (nv_dims.nbDims + 1 != c2_dims.size()) {
+    return false;
+  }
+  for (int i = 0; i < nv_dims.nbDims; ++i) {
+    if (nv_dims.d[i] != c2_dims[i + 1]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+} // namespace
+
+// Upon construction, we build the inference enigne by deserializing from
+// protobuf string. And since we know the input/output blobs, we can do the
+// binding here too.
+TensorRTOp::TensorRTOp(const OperatorDef& operator_def, Workspace* ws)
+    : Operator<CUDAContext>(operator_def, ws),
+      logger_((nvinfer1::ILogger::Severity)(
+          OperatorBase::GetSingleArgument<int>("log_verbosity", 0))),
+      batch_size_(OperatorBase::GetSingleArgument<int>("batch_size", 1)) {
+  {
+    auto engine_string =
+        OperatorBase::GetSingleArgument<std::string>("serialized_engine", "");
+    CAFFE_ENFORCE(!engine_string.empty(), "Empty serialized TensorRT engine!");
+    auto trt_runtime = InferObject(nvinfer1::createInferRuntime(logger_));
+    // TODO(support trt plugin factory)
+    trt_engine_ = InferObject(trt_runtime->deserializeCudaEngine(
+        engine_string.data(), engine_string.size(), nullptr));
+  }
+
+  if(!trt_engine_) {
+    CAFFE_THROW("Cannot deserialize TensorRT engine!");
+  }
+
+  // match and bind the input/output
+  std::unordered_map<std::string, int> inputs;
+  std::unordered_map<std::string, int> outputs;
+  CAFFE_ENFORCE(operator_def.input_size() == InputSize());
+  CAFFE_ENFORCE(operator_def.output_size() == OutputSize());
+  int N = 0;
+  bool first = true;
+  for (int i = 0; i < operator_def.input_size(); ++i) {
+    inputs.emplace(operator_def.input(i), i);
+    // Decide input batch size
+    const auto& input_tensor = Input(i);
+    const auto& tensor_dims = input_tensor.dims();
+    // We need NCHW format
+    CAFFE_ENFORCE(tensor_dims.size() == 4);
+    if (first) {
```

Comment:
nit: I would suggest directly inspect `N` from the first input outside of the for loop to avoid doing `if` `else` in the loop.